modules/core/src/main/scala/org/tessellation/infrastructure/rewards/Rewards.scala

```diff
@@ -1,80 +1,113 @@
 
-      def calculateRewards(
+      def feeDistribution(
+        snapshotOrdinal: SnapshotOrdinal,
+        transactions: SortedSet[transaction.Transaction],
+        facilitators: NonEmptySet[Id]
+      ): F[SortedSet[RewardTransaction]] = {
+
+        val totalFee = transactions.toList
+          .map(_.fee)
+          .foldM(NonNegLong.MinValue) { case (acc, fee) => acc + fee.value }
+          .map(Amount(_))
+          .liftTo[F]
+
+        Random.scalaUtilRandomSeedLong(snapshotOrdinal.value).flatMap { randomizer =>
+          totalFee.flatMap { amount =>
+            regular.distribute(randomizer, facilitators).run(amount).flatMap {
+              case (remaining, rewards) =>
+                if (remaining =!= Amount.empty) {
+                  MonadThrow[F].raiseError[List[(Address, Amount)]](
+                    new RuntimeException(s"Some fees were not distributed {totalFee=${amount.show}, remainingFee=${remaining.show}}")
+                  )
+                } else rewards.pure[F]
+            }
+          }.map(toTransactions)
+        }
+      }
+
+      def mintedDistribution(
         epochProgress: EpochProgress,
         facilitators: NonEmptySet[Id]
       ): F[SortedSet[RewardTransaction]] = {
         val amount = getAmountByEpoch(epochProgress, rewardsPerEpoch)
 
         val programRewardsState = for {
-          stardustCollective <- stardust.distribute(epochProgress, facilitators)
+          stardustCollective <- stardust.distribute()
           softStakingRewards <- softStaking.distribute(epochProgress, facilitators)
-          dtmRewards <- dtm.distribute(epochProgress, facilitators)
+          dtmRewards <- dtm.distribute()
         } yield dtmRewards ++ softStakingRewards ++ stardustCollective
 
         def eitherToF[A](either: Either[ArithmeticException, A]): F[A] = either.liftTo[F]
 
         val allRewardsState = for {
           programRewards <- programRewardsState.mapK[F](lift(eitherToF))
-          regularRewards <- regular.distribute(epochProgress, facilitators)
+          randomizer <- StateT.liftF(Random.scalaUtilRandomSeedLong(epochProgress.coerce))
+          regularRewards <- regular.distribute(randomizer, facilitators)
         } yield programRewards ++ regularRewards
 
         allRewardsState
           .run(amount)
-          .flatTap {
-            case (remaining, _) =>
-              new RuntimeException(s"Some rewards were not distributed {amount=${amount.show}, remainingAmount=${remaining.show}}")
-                .raiseError[F, Unit]
-                .whenA(remaining =!= Amount.empty)
+          .flatMap {
+            case (remaining, rewards) =>
+              if (remaining =!= Amount.empty) {
+                MonadThrow[F].raiseError[List[(Address, Amount)]](
+                  new RuntimeException(s"Some rewards were not distributed {amount=${amount.show}, remainingAmount=${remaining.show}}")
+                )
+              } else rewards.pure[F]
```

Comment:
Please extract lines 95-100 and lines 61-67 into a common function. Also I think that the previous notation was more concise:

```scala
.flatTap {
          case (remaining, _) =>
            new RuntimeException(s"Some rewards were not distributed {amount=${amount.show}, remainingAmount=${remaining.show}}")
              .raiseError[F, Unit]
              .whenA(remaining =!= Amount.empty)
```

Reply:
Cool, I didn't think about using `flatTap` for throwing. Looks nice